### ISSUE TITLE
内部的にノートのテキストをHTMLで保持する

### ISSUE THREAD

**Issue by EbiseLutica** [syuilo/Misskey#7876]
## 目的

古いバージョンのMisskeyやフォーク、そしてその他のMFMでない構文を用いる実装との相互運用性を上げる

## 仕様

* データベースのNote.text, Note.cw などのMFMを挿入する箇所をHTMLを挿入する形に置き換える
* 既存のノートについては、参照されてからMFM→HTML変換をするゆるいマイグレーションを取る
  * 全てを一度に置き換えるのは、大規模なサーバーであるほど負荷が大きい
* MFMの独自構文は `[spin.x.speed=0.2s honi]` → `<span data-type="tag" data-name="spin" data-param-x data-param-speed="0.2s">honi</span>` といった形で置換？
* フロントエンドの対応は以下の2案を提案
  * **A案:** APIでHTMLを返し、フロントエンドにレンダリングさせる
    * **メリット**
      * 非MFM文字列をMFMとしてレンダリングしてしまうことが無くなる
    * **デメリット**
      * フロントエンドの実装負担が増える
  * **B案:** これまで通りMFMを返す
    * **メリット**
      * サードパーティアプリの互換性が保たれる
      * フロントエンドの実装負担が増えない
    * **デメリット**
      * 結局MFMに変換するのであればHTMLで持つ意味が薄れる

**Comment by syuilo** [comment]
全文検索と相性が悪い可能性がありそう

**Comment by EbiseLutica** [comment]
@syuilo たしかに。全文検索にかけるときにプレーンテキスト化するようにするか、プレーンテキスト化したものをplainText, plainCwというフィールドに入れるか。前者は実行時負荷が大きそうで、後者はデータ量が肥大化して大変なことになりそう。

**Comment by Johann150** [comment]
Another advantage of proposal A would be that the clients do not have to newly parse MFM of notes every time (but they have to parse HTML in both proposals). I think it could be a significant performance improvement.
And also, the HTML to MFM conversion could be removed, which has caused multiple issues in the past. For example looking at `git blame src/mfm/from-html.ts` shows at least #2339, #7150, #7839, and #7919 was newly discovered today.

And the better federation with software that does not use MFM would also be something I would personally like.

I think proposal A is good because it removes the unnecessary information loss when converting HTML to MFM, and converting MFM back to HTML.

**Comment by EbiseLutica** [comment]
> Another advantage of proposal A would be that the clients do not have to newly parse MFM of notes every time (but they have to parse HTML in both proposals).

There are many HTML-parser library for several languages around the world, so developer may be able to use them instead implementing MFM-parser.

世の中には様々な言語向けにHTMLパーサーライブラリがあるため、開発者はMFMパーサーを実装する代わりにそれらを利用できます。

**Comment by Johann150** [comment]
Yes, the idea was that many clients are web browsers so they should usually already know how to parse HTML. Also Pleroma, Mastodon etc. all seem to use HTML so it would be more compatible.

**Comment by syuilo** [comment]
全文検索のことがあるから通常のテキストとHTMLとで二重に持つことになると思うけど、相互運用性を高める目的のみにしてはコストが大きい気がする

**Comment by Johann150** [comment]
I do not quite understand how HTML would break full text search. Do you mean something like `some<b>thing</b>` not being found if i search for `something`? But that would also not work if the text was represented by MFM.

Could you give an example where HTML breaks full text search that MFM does not break? I also noticed that I can use the usual SQL `LIKE` placeholders (e.g. `%`) in the full text search so that should also be considered.

**Comment by Johann150** [comment]
related: #8011

**Comment by EbiseLutica** [comment]
> I do not quite understand how HTML would break full text search. Do you mean something like some<b>thing</b> not being found if i search for something? But that would also not work if the text was represented by MFM.

HTMLタグで検索すると、たくさんの無関係の投稿が引っかかるようになってしまう可能性はあります。MFMはユーザーが書いたものですが、HTMLはサーバーが生成しています。これはユーザビリティを損ねる可能性があります。

> 全文検索のことがあるから通常のテキストとHTMLとで二重に持つことになると思うけど、相互運用性を高める目的のみにしてはコストが大きい気がする

現状問題になっている、cat状態で「削除して編集」がうまく動かない不具合の修正のために、APIで生文字列を返すことにも使えそう（これは別issueになってたかな）

あと、ユーザーが入力した生文字列を持っておくのは、法的な開示などにも対応しやすくなるメリットもあるし、将来的に変換方法を変えたとしてもマイグレしやすくなりそうかなーっておもう

**Comment by EbiseLutica** [comment]
ただこれ、Misskeyはユーザー名や自己紹介にまで装飾を使えてしまうので、MFMを利用し得る全てのデータ構造に対して生データを持てという話になってくる可能性があり、難しいかも。
